pkg/streamer/reader.go

```diff
@@ -103,9 +117,206 @@ func (r *BinlogReader) checkRelayPos(pos mysql.Position) error {
+	for _, uuid := range r.uuids {
+		filename := path.Join(r.cfg.RelayDir, uuid, utils.MetaFilename)
+		fd, err := os.Open(filename)
+		if err != nil {
+			return "", terror.ErrRelayLoadMetaData.Delegate(err)
+		}
+		defer fd.Close()
+
+		var meta Meta
+		_, err = toml.DecodeReader(fd, &meta)
+		if err != nil {
+			return "", terror.ErrRelayLoadMetaData.Delegate(err)
+		}
+
+		gs, err := mysql.ParseGTIDSet(r.cfg.Flavor, meta.BinlogGTID)
+		if err != nil {
+			return "", terror.ErrRelayLoadMetaData.Delegate(err)
+		}
+		if gs.Contain(gset) {
+			r.tctx.L().Info("get uuid subdir by gtid", zap.Stringer("GTID Set", gset), zap.String("uuid", uuid))
+			return uuid, nil
+		}
+	}
+
+	// use memory meta
+	if len(relayMeta.UUID) > 0 {
+		gs, err := mysql.ParseGTIDSet(r.cfg.Flavor, relayMeta.BinlogGTID)
+		if err != nil {
+			return "", terror.ErrRelayLoadMetaData.Delegate(err)
+		}
+		if gs.Contain(gset) {
+			r.tctx.L().Info("get uuid subdir by gtid", zap.Stringer("GTID Set", gset), zap.String("uuid", relayMeta.UUID))
+			return relayMeta.UUID, nil
+		}
+	}
+	return "", terror.ErrNoUUIDDirMatchGTID.Generate(gset.String())
+}
+
+// GetFilePosByGTID tries to get Pos by GTID for file
+func (r *BinlogReader) GetFilePosByGTID(ctx context.Context, filePath string, gset mysql.GTIDSet, oldest bool) (uint32, error) {
+	ctx2, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	s := newLocalStreamer()
+	parser := replication.NewBinlogParser()
+	parser.SetVerifyChecksum(true)
+	parser.SetUseDecimal(true)
+	if r.cfg.Timezone != nil {
+		parser.SetTimestampStringLocation(r.cfg.Timezone)
+	}
+
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+
+		offset := int64(4)
+		onEventFunc := func(e *replication.BinlogEvent) error {
+			offset = int64(e.Header.LogPos)
+
+			select {
+			case s.ch <- e:
+			case <-ctx2.Done():
+			}
+			return nil
+		}
+
+		err := parser.ParseFile(filePath, offset, onEventFunc)
+		if err != nil {
+			r.tctx.L().Error("parse file stopped", zap.Error(err))
+			select {
+			case s.ech <- err:
+			case <-ctx2.Done():
+			}
+		} else {
+			// reach end of file
+			select {
+			case s.ch <- nil:
+			case <-ctx2.Done():
+			}
+		}
+	}()
+
+	lastPos := uint32(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return 0, nil
+		default:
+		}
+
+		e, err := s.GetEvent(ctx)
+		if err != nil {
+			return 0, err
+		}
+		// gset is the last txn in file
+		if e == nil {
+			return lastPos, nil
+		}
+		switch ev := e.Event.(type) {
+		case *replication.PreviousGTIDsEvent:
+			// nil previous gtid event
+			if len(ev.GTIDSets) == 0 {
+				lastPos = e.Header.LogPos
+				continue
+			}
+			gs, err := mysql.ParseGTIDSet(r.cfg.Flavor, ev.GTIDSets)
+			if err != nil {
+				return 0, err
+			}
+			// if PreviousGITDsEvent contain the gset, go to previous file
+			if gs.Contain(gset) {
+				// if it's the oldest file and gs equals gset, continue
+				if oldest && gset.Contain(gs) {
+					lastPos = e.Header.LogPos
+					continue
+				}
```

Comment:
mysql.000001: [..., gtid, query, heartbeat, ..., rotate]
mysql.000002: [format, previous_gtid]
We cant start from mysql.000001's heartbeat event or mysql.000002's previous_gtid event.
What are your opinions? @lance6716 @csuzhangxc